File: backend/internal/model/payload/period.go
package payload

type PeriodMonthStartedPayload struct {
	PeriodStart string `json:"period_start"`
}

func (p PeriodMonthStartedPayload) Validate() error {
	var errs []string
	if p.PeriodStart == "" {
		errs = append(errs, "period_start is required")
	}

	return joinErrors(errs)
}

type PeriodMonthClosedPayload struct {
	ClosingId int64  `json:"closing_id"`
	ClosedAt  string `json:"closed_at"`
}

func (p PeriodMonthClosedPayload) Validate() error {
	var errs []string

	if p.ClosingId == 0 {
		errs = append(errs, "closing_id is required")
	}

	if p.ClosedAt == "" {
		errs = append(errs, "closed_at is required")
	}

	return joinErrors(errs)
}

type PeriodMonthReopenedPayload struct {
	ClosingId  int64  `json:"closing_id"`
	Reason     string `json:"reason"`
	ReopenedAt string `json:"reopened_at"`
}

func (p PeriodMonthReopenedPayload) Validate() error {
	var errs []string
	if p.ClosingId == 0 {
		errs = append(errs, "closing_id is required")
	}

	if p.ReopenedAt == "" {
		errs = append(errs, "reopened_at is required")
	}

	if p.Reason == "" {
		errs = append(errs, "reason is required")
	}

	return joinErrors(errs)
}

type PeriodAnnualStartedPayload struct {
	Year int `json:"year"`
}

func (p PeriodAnnualStartedPayload) Validate() error {
	var errs []string
	if p.Year == 0 {
		errs = append(errs, "year is required")
	}
	return joinErrors(errs)
}

type PeriodAnnualClosedPayload struct {
	ClosingId int64  `json:"closing_id"`
	ClosedAt  string `json:"closed_at"`
}

func (p PeriodAnnualClosedPayload) Validate() error {
	var errs []string
	if p.ClosingId == 0 {
		errs = append(errs, "closing_id is required")
	}

	if p.ClosedAt == "" {
		errs = append(errs, "closed_at is required")
	}
	return joinErrors(errs)
}

type PeriodAnnualReopenedPayload struct {
	ClosingId  int64  `json:"closing_id"`
	Reason     string `json:"reason"`
	ReopenedAt string `json:"reopened_at"`
}

func (p PeriodAnnualReopenedPayload) Validate() error {
	var errs []string
	if p.ClosingId == 0 {
		errs = append(errs, "closing_id is required")
	}

	if p.ReopenedAt == "" {
		errs = append(errs, "reopened_at is required")
	}

	if p.Reason == "" {
		errs = append(errs, "reason is required")
	}

	return joinErrors(errs)
}

File: frontside/src/api/period.ts
import { apiFetch } from './http';
import { getAggregateVersion, appendEvent } from './aggregate';
import type {
  PeriodClosing,
  PeriodType,
  PeriodMonthStartedPayload,
  PeriodMonthClosedPayload,
  PeriodMonthReopenedPayload,
  PeriodAnnualStartedPayload,
  PeriodAnnualClosedPayload,
  PeriodAnnualReopenedPayload,
} from '../types/period';
import type { PaginatedResponse } from '../types/pagination';

export interface GetPeriodPagedParams {
  page:     number;
  pageSize: number;
}

export const getPeriodPaged = async (
  periodType: PeriodType,
  params: GetPeriodPagedParams,
): Promise<PaginatedResponse<PeriodClosing>> => {
  const qs = new URLSearchParams({
    page:      String(params.page),
    page_size: String(params.pageSize),
  });

  const response = await apiFetch(`/api/period/${periodType}?${qs}`);
  if (!response.ok) {
    const problem = await response.json();
    throw new Error(problem.detail ?? problem.title ?? '查詢失敗');
  }
  return response.json();
};

// ── 月結事件 ──────────────────────────────────

export const startMonthPeriod = async (payload: PeriodMonthStartedPayload): Promise<void> => {
  const version = await getAggregateVersion('ACCOUNT', '');
  await appendEvent({
    aggregate_type:   'ACCOUNT',
    aggregate_id:     '',
    expected_version: version,
    event_type:       'period.month_started',
    payload,
    metadata:         {},
  });
};

export const closeMonthPeriod = async (payload: PeriodMonthClosedPayload): Promise<void> => {
  const version = await getAggregateVersion('ACCOUNT', '');
  await appendEvent({
    aggregate_type:   'ACCOUNT',
    aggregate_id:     '',
    expected_version: version,
    event_type:       'period.month_closed',
    payload,
    metadata:         {},
  });
};

export const reopenMonthPeriod = async (payload: PeriodMonthReopenedPayload): Promise<void> => {
  const version = await getAggregateVersion('ACCOUNT', '');
  await appendEvent({
    aggregate_type:   'ACCOUNT',
    aggregate_id:     '',
    expected_version: version,
    event_type:       'period.month_reopened',
    payload,
    metadata:         {},
  });
};

// ── 年結事件 ──────────────────────────────────

export const startAnnualPeriod = async (payload: PeriodAnnualStartedPayload): Promise<void> => {
  const version = await getAggregateVersion('ACCOUNT', '');
  await appendEvent({
    aggregate_type:   'ACCOUNT',
    aggregate_id:     '',
    expected_version: version,
    event_type:       'period.annual_started',
    payload,
    metadata:         {},
  });
};

export const closeAnnualPeriod = async (payload: PeriodAnnualClosedPayload): Promise<void> => {
  const version = await getAggregateVersion('ACCOUNT', '');
  await appendEvent({
    aggregate_type:   'ACCOUNT',
    aggregate_id:     '',
    expected_version: version,
    event_type:       'period.annual_closed',
    payload,
    metadata:         {},
  });
};

export const reopenAnnualPeriod = async (payload: PeriodAnnualReopenedPayload): Promise<void> => {
  const version = await getAggregateVersion('ACCOUNT', '');
  await appendEvent({
    aggregate_type:   'ACCOUNT',
    aggregate_id:     '',
    expected_version: version,
    event_type:       'period.annual_reopened',
    payload,
    metadata:         {},
  });
};
